Modules/_json.c

```diff
@@ -1503,6 +1503,13 @@ encoder_listencode_obj(PyEncoderObject *s, _PyAccu *acc,
             }
         }
         newobj = PyObject_CallFunctionObjArgs(s->defaultfn, obj, NULL);
+        if (obj != NULL && newobj != NULL && newobj->ob_type == obj->ob_type) {
```

Comment:
Wouldn't PyObject_IsInstance raise false positives for our condition? For this fix, we would want the exact types to match, isn't it?